### PR TITLE
Sweep generic mnemonic's

### PR DESCRIPTION
**Pull request by lacksfish**
Please review and discuss.

I'm not sure where we should put the button yet.

### CODE REVIEW

File: src/js/modules/core/services/sweeper/sweeper-service.factory.js [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+                var outputValue = totalValue - approxFee;
+
+                if (outputValue > 547) {
+                    rawTransaction.addOutput(options.recipient, outputValue);
+
+                    angular.forEach(inputs, function (input, index) {
+                        var currentPrivKey = resultUTXOs[input.address].priv_key;
+
+                        // Set hashType depending on network
+                        var hashType = bitcoinJS.Transaction.SIGHASH_ALL | (options.network.toLowerCase() == 'bcc' ? bitcoinJS.Transaction.SIGHASH_BITCOINCASHBIP143 : 0);
```
Comment (afk11) [inline src/js/modules/core/services/sweeper/sweeper-service.factory.js]: var hashType = bitcoinJS.Transaction.SIGHASH_ALL
if (options.network.toLowerCase() == 'bcc') {
    hashType |= bitcoinJS.Transaction.SIGHASH_BITCOINCASHBIP143
}

File: src/js/modules/core/services/sweeper/sweeper-service.factory.js [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+                        totalValue: totalValue
+                    }
+                } else {
+                    return false;
+                }
+            });
+        }
+
+        function hdDiscover(root, batchSize, accountBatchSize, testnet, accountIdx, UTXOs) {
+            testnet = testnet || 0;
```
Comment (afk11) [inline src/js/modules/core/services/sweeper/sweeper-service.factory.js]: The testnet parameter is passed as a boolean in some places, but here it defaults to 0, and it's only used as a string in the paths..

File: src/js/modules/core/services/sweeper/sweeper-service.factory.js [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        function genericHdNodeSweep(root, options) {
+            return launchService.getAccountInfo()
+                .then(function (accountInfo) {
+                    bitcoinDataClient = new BlocktrailBitcoinService({
+                        apiKey: accountInfo.api_key,
+                        apiSecret: accountInfo.api_secret,
+                        network: options.network,
+                        testnet: options.testnet
+                    });
+                    return hdDiscover(root, options.batchSize, options.accountBatchSize, !!options.testnet)
```
Comment (afk11) [inline src/js/modules/core/services/sweeper/sweeper-service.factory.js]: `!!options.testnet` is where it's passed as a boolean..

File: src/js/modules/core/services/sweeper/sweeper-service.factory.js (hunk 2 above)
Comment (afk11) [inline src/js/modules/core/services/sweeper/sweeper-service.factory.js]:
```
-            testnet = testnet || 0;
+           var testnetIdx = testnet ? 1 : 0
```

File: src/js/modules/core/services/sweeper/sweeper-service.factory.js [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+            }
+
+            var entropy = new blocktrailSDK.Buffer(seed, 'hex');
+            var root = bitcoinJS.HDNode.fromSeedBuffer(entropy);
+
+            return genericHdNodeSweep(root, options)
+                .then(function (resultUTXOs) {
+                    return signAndSweep(resultUTXOs, options).then(function (result) {
+                        return result;
+                    })
```
Comment (afk11) [inline src/js/modules/core/services/sweeper/sweeper-service.factory.js]: Remainants of debugging code? `return signAndSweep(reusltUTXOs, options);`

File: src/js/controllers/WalletControllers.js [hunk 5]
```diff
@@ -1,8 +1,15 @@
 angular.module('blocktrail.wallet')
-    .controller('WalletCtrl', function($q, $log, $scope, $state, $rootScope, $interval, storageService, sdkService, Wallet,
+    .controller('WalletCtrl', function($q, $log, $scope, $state, $rootScope, $interval, storageService, sdkService, Wallet, $modal,
                                        Contacts, CONFIG, settingsService, setupService, $timeout, launchService, blocktrailLocalisation,
                                        dialogService, $http, $translate, buyBTCService, Currencies, AppVersionService, $filter) {
 
+        $scope.test = function () {
```
Comment (afk11) [inline src/js/controllers/WalletControllers.js]: Give this a name :)

File: src/js/modules/core/services/sweeper/sweeper-service.factory.js [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+        function batchDiscoverUTXOs(parentNode, batchLoopCount, batchSize) {
+            var batchDataObject = batchDeriveAddressesAndPrivKeys(parentNode, batchLoopCount * batchSize, batchSize);
+            var addresses = Object.keys(batchDataObject);
+
+            return bitcoinDataClient.batchAddressHasTransactions(addresses).then(function (success) {
+                if (!success) {
+                    return false;
+                } else {
+                    return bitcoinDataClient.getBatchUnspentOutputs(addresses).then(function (result) {
```
Comment (afk11) [inline src/js/modules/core/services/sweeper/sweeper-service.factory.js]: Maybe move the handling part here down, to be run when promise returned by the present function returns the batch of utxos?

File: src/js/modules/core/services/sweeper/sweeper-service.factory.js [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+                        testnet: options.testnet
+                    });
+                    return hdDiscover(root, options.batchSize, options.accountBatchSize, !!options.testnet)
+                });
+        }
+
+        function genericSweep(mnemonic, options) {
+            var seed = bip39.mnemonicToSeed(mnemonic);
+
+            if(! bip39.validateMnemonic(mnemonic)) {
```
Comment (afk11) [inline src/js/modules/core/services/sweeper/sweeper-service.factory.js]: Validate the mnemonic first

File: src/js/modules/core/services/sweeper/sweeper-service.factory.js (hunk 1 above)
Comment (lacksfish) [inline src/js/modules/core/services/sweeper/sweeper-service.factory.js]: Nice, thanks.

File: src/js/controllers/WalletControllers.js (hunk 5 above)
Comment (lacksfish) [inline src/js/controllers/WalletControllers.js]: It shouldn't stay there anyways.

File: src/js/modules/core/services/sweeper/sweeper-service.factory.js (hunk 7 above)
Comment (lacksfish) [inline src/js/modules/core/services/sweeper/sweeper-service.factory.js]: Sure, silly me.

File: src/js/modules/core/services/sweeper/sweeper-service.factory.js (hunk 4 above)
Comment (lacksfish) [inline src/js/modules/core/services/sweeper/sweeper-service.factory.js]: Yes, thanks.

File: src/js/modules/core/services/sweeper/sweeper-service.factory.js (hunk 2 above)
Comment (lacksfish) [inline src/js/modules/core/services/sweeper/sweeper-service.factory.js]: Great, thanks. I've missed that. :)

File: src/js/modules/core/services/sweeper/sweeper-service.factory.js (hunk 6 above)
Comment (lacksfish) [inline src/js/modules/core/services/sweeper/sweeper-service.factory.js]: Thanks.

File: src/js/modules/core/services/sweeper/sweeper-service.factory.js (hunk 3 above)
Comment (lacksfish) [inline src/js/modules/core/services/sweeper/sweeper-service.factory.js]: So getting rid of the the double-not, I'd say, later on `testnet ? 1 : 0` evaluates it anyways

File: src/js/modules/core/services/sweeper/sweeper-service.factory.js [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+        function signAndSweep(resultUTXOs, options) {
+            // Create raw transaction
+            var inputs = [];
+            var totalValue = 0;
+            var addresses = Object.keys(resultUTXOs);
+            var rawTransaction = new bitcoinJS.TransactionBuilder();
+            var hashType = bitcoinJS.Transaction.SIGHASH_ALL;
+
+            // If Bitcoin Cash sweep
+            if (options.network.toLowerCase() == 'bcc') {
```
Comment (rubensayshi) [inline src/js/modules/core/services/sweeper/sweeper-service.factory.js]: use `===` when types are equal on both sides

File: src/js/app.js [hunk 9]
```diff
@@ -15,6 +15,7 @@ var blocktrail = angular.module('blocktrail.wallet', [
     'blocktrail.config',
     'blocktrail.core',
     'blocktrail.setup',
+    'blocktrail.tools', // TODO: Rename
```
Comment (lacksfish) [inline src/js/app.js]: Should've removed that, sorry.

File: src/js/modules/wallet/controllers/sweep-coins-modal/sweep-coins-modal.ctrl.js [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        $scope.sweepData = {
+            rawTx: null,
+            feePaid: null,
+            inputCount: null,
+            totalValue: null
+        };
+
+        var options = {
+            batchSize: 50,
+            accountBatchSize: 1,
```
Comment (afk11) [inline src/js/modules/wallet/controllers/sweep-coins-modal/sweep-coins-modal.ctrl.js]: This can vary when it's Mycelium I guess?

Comment (lacksfish) [inline src/js/modules/wallet/controllers/sweep-coins-modal/sweep-coins-modal.ctrl.js]: Yes, accountBatchSize is for how many accounts to check in one cycle.

Comment (lacksfish) [inline src/js/modules/wallet/controllers/sweep-coins-modal/sweep-coins-modal.ctrl.js]: Which should ever only be one account (account index 0) in blockchain.infos case

**Pull request merged**